src/issuance.rs

```diff
@@ -121,35 +115,39 @@ impl IssueAction {
     ///
     /// This function may return an error in any of the following cases:
     ///
-    /// * `IssueActionIncorrectAssetBase`: If the asset type of any note in the `IssueAction` is
-    ///   not equal to the asset type of the first note.
-    ///
     /// * `ValueSumOverflow`: If the total amount value of all notes in the `IssueAction` overflows.
     ///
     /// * `IssueBundleIkMismatchAssetBase`: If the provided `ik` is not used to derive the
     ///   `AssetBase` for **all** internal notes.
+    ///
+    /// * `IssueActionWithoutNoteNotFinalized`:If the `IssueAction` contains no note and is not finalized.
     fn verify_supply(&self, ik: &IssuanceValidatingKey) -> Result<(AssetBase, AssetSupply), Error> {
+        let issue_action_asset = AssetBase::derive(ik, &self.asset_desc);
+
         // Calculate the value of the asset as a sum of values of all its notes
-        // and ensure all note types are equal
-        let (asset, value_sum) = self.notes.iter().try_fold(
-            (self.notes().head.asset(), ValueSum::zero()),
-            |(asset, value_sum), &note| {
+        // and ensure all note types are equal the asset derived from asset_desc and ik.
+        let value_sum = self
+            .notes
+            .iter()
+            .try_fold(ValueSum::zero(), |value_sum, &note| {
                 // All assets should have the same `AssetBase`
                 note.asset()
-                    .eq(&asset)
+                    .eq(&issue_action_asset)
                     .then(|| ())
-                    .ok_or(IssueActionIncorrectAssetBase)?;
+                    .ok_or(IssueBundleIkMismatchAssetBase)?;
 
                 // The total amount should not overflow
-                Ok((asset, (value_sum + note.value()).ok_or(ValueSumOverflow)?))
-            },
-        )?;
+                (value_sum + note.value()).ok_or(ValueSumOverflow)
+            })?;
 
-        // Return the asset and its supply (or an error if the asset was not properly derived)
-        asset
-            .eq(&AssetBase::derive(ik, &self.asset_desc))
-            .then(|| Ok((asset, AssetSupply::new(value_sum, self.is_finalized()))))
-            .ok_or(IssueBundleIkMismatchAssetBase)?
+        if self.notes.is_empty() && !self.is_finalized() {
+            Err(IssueActionWithoutNoteNotFinalized)
+        } else {
+            Ok((
+                issue_action_asset,
+                AssetSupply::new(value_sum, self.is_finalized()),
+            ))
+        }
```

Comment:
let's split the error from the ok:
```rust
    fn verify_supply(&self, ik: &IssuanceValidatingKey) -> Result<(AssetBase, AssetSupply), Error> {
        if self.notes.is_empty() && !self.is_finalized() {
            return Err(IssueActionWithoutNoteNotFinalized);
        }        
               
        let issue_asset = AssetBase::derive(ik, &self.asset_desc);
        
        // Calculate the value of the asset as a sum of values of all its notes
        // and ensure all note types are equal the asset derived from asset_desc and ik.
        let value_sum = self
            .notes
            .iter()
            .try_fold(ValueSum::zero(), |value_sum, &note| {
                // All assets shouldbe derived correctly
                note.asset()
                    .eq(&issue_asset)
                    .then(|| ())
                    .ok_or(IssueBundleIkMismatchAssetBase)?;

                // The total amount should not overflow
                (value_sum + note.value()).ok_or(ValueSumOverflow)
            })?;
        
        Ok((
            issue_asset,
            AssetSupply::new(value_sum, self.is_finalized()),
        ))
    }
``` 
note the name change for `issue_action_asset`

Reply:
Done